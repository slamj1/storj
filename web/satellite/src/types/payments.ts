// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all payments-related functionality
 */
export interface PaymentsApi {
    /**
     * Try to set up a payment account
     *
     * @throws Error
     */
    setupAccount(): Promise<void>;

    /**
     * Get account balance
     *
     * @returns balance in cents
     * @throws Error
     */
    getBalance(): Promise<number>;

    /**
     * projectsUsagesAndCharges returns usage and how much money current user will be charged for each project which he owns.
     */
    projectsUsageAndCharges(since: Date, before: Date): Promise<ProjectUsageAndCharges[]>;

    /**
     * Add credit card
     * @param token - stripe token used to add a credit card as a payment method
     * @throws Error
     */
    addCreditCard(token: string): Promise<void>;

    /**
     * Detach credit card from payment account.
     * @param cardId
     * @throws Error
     */
    removeCreditCard(cardId: string): Promise<void>;

    /**
     * Get list of user`s credit cards
     *
     * @returns list of credit cards
     * @throws Error
     */
    listCreditCards(): Promise<CreditCard[]>;

    /**
     * Make credit card default
     * @param cardId
     * @throws Error
     */
    makeCreditCardDefault(cardId: string): Promise<void>;

    /**
     * Returns a list of invoices, transactions and all others billing history items for payment account.
     *
     * @returns list of billing history items
     * @throws Error
     */
    billingHistory(): Promise<BillingHistoryItem[]>;

    /**
     * Creates token transaction in CoinPayments
     *
     * @param amount
     * @throws Error
     */
    makeTokenDeposit(amount: number): Promise<TokenDeposit>;
}

export class CreditCard {
    public isSelected: boolean = false;

    constructor(
        public id: string = '',
        public expMonth: number = 0,
        public expYear: number = 0,
        public brand: string = '',
        public last4: string = '0000',
        public isDefault: boolean = false,
    ) {}
}

export class PaymentAmountOption {
    public constructor(
        public value: number,
        public label: string = '',
    ) {}
}

// BillingHistoryItem holds all public information about billing history line.
export class BillingHistoryItem {
    public constructor(
        public readonly id: string = '',
        public readonly description: string = '',
        public readonly amount: number = 0,
        public readonly received: number = 0,
        public readonly status: string = '',
        public readonly link: string = '',
        public readonly start: Date = new Date(),
        public readonly end: Date = new Date(),
        public readonly type: BillingHistoryItemType = BillingHistoryItemType.Invoice,
    ) {}

    public get quantity(): Amount {
        if (this.type === BillingHistoryItemType.Transaction) {
            return new Amount('USD $', this.amountDollars(this.amount), this.amountDollars(this.received));
        }

        return new Amount('USD $', this.amountDollars(this.amount));
    }

    public get formattedStatus(): string {
        return this.status.charAt(0).toUpperCase() + this.status.substring(1);
    }

    private amountDollars(amount): number {
        return amount / 100;
    }

    public downloadLinkHtml(): string {
        if (!this.link) {
            return '';
        }

        const downloadLabel = this.type === BillingHistoryItemType.Transaction ? 'Checkout' : 'PDF';

        return `<a class="download-link" target="_blank" href="${this.link}">${downloadLabel}</a>`;
    }
}

/**
 * BillingHistoryItemType indicates type of billing history item.
  */
export enum BillingHistoryItemType {
    // Invoice is a Stripe invoice billing item.
    Invoice = 0,
    // Transaction is a Coinpayments transaction billing item.
    Transaction = 1,
    // Charge is a credit card charge billing item.
    Charge = 2,
}

/**
 * BillingHistoryStatusType indicates status of billing history item.
 */
export enum BillingHistoryItemStatus {
    /**
     * Status showed if transaction successfully completed.
     */
    Completed = 'completed',

    /**
     * Status showed if transaction successfully paid.
     */
    Paid = 'paid',

    /**
     * Status showed if transaction is pending.
     */
    Pending = 'pending',
}

/**
 * TokenDeposit holds public information about token deposit.
 */
export class TokenDeposit {
    constructor(
        public amount: number,
        public address: string,
        public link: string,
    ) {}
}

/**
 * Amount holds information for displaying billing item payment.
 */
class Amount {
    public constructor(
        public currency: string = '',
        public total: number = 0,
        public received: number = 0,
    ) {}
}

/**
 * ProjectUsageAndCharges shows usage and how much money current project will charge in the end of the month.
  */
export class ProjectUsageAndCharges {
    public constructor(
        public since: Date = new Date(),
        public before: Date = new Date(),
        public egress: number = 0,
        public storage: number = 0,
        public objectCount: number = 0,
        public projectId: string = '',
        // storage shows how much cents we should pay for storing GB*Hrs.
        public storagePrice: number = 0,
        // egress shows how many cents we should pay for Egress.
        public egressPrice: number = 0,
        // objectCount shows how many cents we should pay for objects count.
        public objectPrice: number = 0) {}

    /**
     * summary returns total price for a project in cents.
     */
    public summary(): number {
        return this.storagePrice + this.egressPrice + this.objectPrice;
    }
}

/**
 * Holds start and end dates.
 */
export class DateRange {
    public startDate: Date = new Date();
    public endDate: Date = new Date();

    public constructor(startDate: Date, endDate: Date) {
        this.startDate = startDate;
        this.endDate = endDate;
    }
}
