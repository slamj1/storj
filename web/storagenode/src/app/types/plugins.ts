// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { Telemetry } from '@/app/telemetry/telemetry';

declare module 'vue/types/vue' {
    interface Vue {
        $telemetry: Telemetry;
    }
}
