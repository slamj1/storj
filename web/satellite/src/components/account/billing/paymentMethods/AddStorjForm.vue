// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="add-storj-area">
        <div class="add-storj-area__selection-container">
            <p class="add-storj-area__selection-container__label">Deposit STORJ Tokens via Coin Payments</p>
            <TokenDepositSelection
                class="add-storj-area__selection-container__form"
                @onChangeTokenValue="onChangeTokenValue"
            />
        </div>
        <div class="add-storj-area__submit-area">
            <img
                v-if="isLoading"
                class="loading-image"
                src="@/../static/images/account/billing/loading.gif"
                alt="loading gif"
            >
            <VButton
                class="confirm-add-storj-button"
                label="Continue to Coin Payments"
                width="251px"
                height="48px"
                :on-press="onConfirmAddSTORJ"
                :is-disabled="isLoading"
            />
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import TokenDepositSelection from '@/components/account/billing/paymentMethods/TokenDepositSelection.vue';
import VButton from '@/components/common/VButton.vue';

import { PAYMENTS_ACTIONS } from '@/store/modules/payments';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';
import { ProjectOwning } from '@/utils/projectOwning';

const {
    MAKE_TOKEN_DEPOSIT,
    GET_BILLING_HISTORY,
} = PAYMENTS_ACTIONS;

@Component({
    components: {
        TokenDepositSelection,
        VButton,
    },
})
export default class AddStorjForm extends Vue {
    private readonly DEFAULT_TOKEN_DEPOSIT_VALUE = 50; // in dollars.
    private readonly MAX_TOKEN_AMOUNT = 1000000; // in dollars.
    private tokenDepositValue: number = this.DEFAULT_TOKEN_DEPOSIT_VALUE;

    @Prop({ default: false })
    public readonly isLoading: boolean;

    /**
     * onConfirmAddSTORJ checks if amount is valid and if so process token.
     * payment and return state to default
     */
    public async onConfirmAddSTORJ(): Promise<void> {
        this.$emit('toggleIsLoading');

        if (!this.isDepositValueValid) return;

        try {
            const tokenResponse = await this.$store.dispatch(MAKE_TOKEN_DEPOSIT, this.tokenDepositValue * 100);
            await this.$notify.success(`Successfully created new deposit transaction! \nAddress:${tokenResponse.address} \nAmount:${tokenResponse.amount}`);
            const depositWindow = window.open(tokenResponse.link, '_blank');
            if (depositWindow) {
                depositWindow.focus();
            }
        } catch (error) {
            await this.$notify.error(error.message);
            this.$emit('toggleIsLoading');
        }

        this.$segment.track(SegmentEvent.PAYMENT_METHOD_ADDED, {
            project_id: this.$store.getters.selectedProject.id,
        });

        this.tokenDepositValue = this.DEFAULT_TOKEN_DEPOSIT_VALUE;
        try {
            await this.$store.dispatch(GET_BILLING_HISTORY);
        } catch (error) {
            await this.$notify.error(error.message);
            this.$emit('toggleIsLoading');
        }

        this.$emit('toggleIsLoading');
        this.$emit('cancel');
    }

    /**
     * Event for changing token deposit value.
     */
    public onChangeTokenValue(value: number): void {
        this.tokenDepositValue = value;
    }

    /**
     * Indicates if user has own project.
     */
    private get userHasOwnProject(): boolean {
        return new ProjectOwning(this.$store).userHasOwnProject();
    }

    /**
     * Indicates if deposit value is valid.
     */
    private get isDepositValueValid(): boolean {
        switch (true) {
            case (this.tokenDepositValue < this.DEFAULT_TOKEN_DEPOSIT_VALUE || this.tokenDepositValue >= this.MAX_TOKEN_AMOUNT) && !this.userHasOwnProject:
                this.$notify.error('First deposit amount must be more than $50 and less than $1000000');
                this.setDefault();

                return false;
            case this.tokenDepositValue >= this.MAX_TOKEN_AMOUNT || this.tokenDepositValue === 0:
                this.$notify.error('Deposit amount must be more than $0 and less than $1000000');
                this.setDefault();

                return false;
            default:
                return true;
        }
    }

    /**
     * Sets adding payment method state to default.
     */
    private setDefault(): void {
        this.tokenDepositValue = this.DEFAULT_TOKEN_DEPOSIT_VALUE;
        this.$emit('toggleIsLoading');
        this.$emit('cancel');
    }
}
</script>

<style scoped lang="scss">
    p {
        margin: 0;
    }

    .add-storj-area {
        margin-top: 44px;
        font-family: 'font_regular', sans-serif;
        display: flex;
        max-height: 52px;
        justify-content: space-between;
        align-items: center;

        &__selection-container {
            display: flex;
            align-items: center;

            &__label {
                margin-right: 30px;
                max-width: 215px;
            }

            &__form {
                width: 60%;
            }
        }

        &__submit-area {
            display: flex;
            align-items: center;
            min-width: 135px;
        }
    }

    .loading-image {
        width: 18px;
        height: 18px;
        margin-right: 5px;
    }
</style>
