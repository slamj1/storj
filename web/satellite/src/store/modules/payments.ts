// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import {
    BillingHistoryItem,
    BillingHistoryItemStatus,
    BillingHistoryItemType,
    CreditCard,
    DateRange,
    PaymentsApi,
    ProjectUsageAndCharges,
    TokenDeposit,
} from '@/types/payments';

export const PAYMENTS_MUTATIONS = {
    SET_BALANCE: 'SET_BALANCE',
    SET_CREDIT_CARDS: 'SET_CREDIT_CARDS',
    SET_DATE: 'SET_DATE',
    CLEAR: 'CLEAR_PAYMENT_INFO',
    UPDATE_CARDS_SELECTION: 'UPDATE_CARDS_SELECTION',
    UPDATE_CARDS_DEFAULT: 'UPDATE_CARDS_DEFAULT',
    SET_BILLING_HISTORY: 'SET_BILLING_HISTORY',
    SET_PROJECT_USAGE_AND_CHARGES: 'SET_PROJECT_USAGE_AND_CHARGES',
    SET_CURRENT_ROLLUP_PRICE: 'SET_CURRENT_ROLLUP_PRICE',
    SET_PREVIOUS_ROLLUP_PRICE: 'SET_PREVIOUS_ROLLUP_PRICE',
    SET_PRICE_SUMMARY: 'SET_PRICE_SUMMARY',
};

export const PAYMENTS_ACTIONS = {
    GET_BALANCE: 'getBalance',
    SETUP_ACCOUNT: 'setupAccount',
    GET_CREDIT_CARDS: 'getCreditCards',
    ADD_CREDIT_CARD: 'addCreditCard',
    CLEAR_PAYMENT_INFO: 'clearPaymentInfo',
    TOGGLE_CARD_SELECTION: 'toggleCardSelection',
    CLEAR_CARDS_SELECTION: 'clearCardsSelection',
    MAKE_CARD_DEFAULT: 'makeCardDefault',
    REMOVE_CARD: 'removeCard',
    GET_BILLING_HISTORY: 'getBillingHistory',
    MAKE_TOKEN_DEPOSIT: 'makeTokenDeposit',
    GET_PROJECT_USAGE_AND_CHARGES: 'getProjectUsageAndCharges',
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP: 'getProjectUsageAndChargesCurrentRollup',
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP: 'getProjectUsageAndChargesPreviousRollup',
};

const {
    SET_BALANCE,
    SET_CREDIT_CARDS,
    SET_DATE,
    CLEAR,
    UPDATE_CARDS_SELECTION,
    UPDATE_CARDS_DEFAULT,
    SET_BILLING_HISTORY,
    SET_PROJECT_USAGE_AND_CHARGES,
    SET_CURRENT_ROLLUP_PRICE,
    SET_PREVIOUS_ROLLUP_PRICE,
    SET_PRICE_SUMMARY,
} = PAYMENTS_MUTATIONS;

const {
    GET_BALANCE,
    SETUP_ACCOUNT,
    GET_CREDIT_CARDS,
    ADD_CREDIT_CARD,
    TOGGLE_CARD_SELECTION,
    CLEAR_CARDS_SELECTION,
    CLEAR_PAYMENT_INFO,
    MAKE_CARD_DEFAULT,
    REMOVE_CARD,
    GET_BILLING_HISTORY,
    MAKE_TOKEN_DEPOSIT,
    GET_PROJECT_USAGE_AND_CHARGES,
    GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP,
    GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP,
} = PAYMENTS_ACTIONS;

export class PaymentsState {
    /**
     * balance stores in cents
     */
    public balance: number = 0;
    public creditCards: CreditCard[] = [];
    public billingHistory: BillingHistoryItem[] = [];
    public usageAndCharges: ProjectUsageAndCharges[] = [];
    public priceSummary: number = 0;
    public currentRollupPrice: number = 0;
    public previousRollupPrice: number = 0;
    public startDate: Date = new Date();
    public endDate: Date = new Date();
}

/**
 * creates payments module with all dependencies
 *
 * @param api - payments api
 */
export function makePaymentsModule(api: PaymentsApi): StoreModule<PaymentsState> {
    return {
        state: new PaymentsState(),
        mutations: {
            [SET_BALANCE](state: PaymentsState, balance: number): void {
                state.balance = balance;
            },
            [SET_CREDIT_CARDS](state: PaymentsState, creditCards: CreditCard[]): void {
                state.creditCards = creditCards;
            },
            [SET_DATE](state: PaymentsState, dateRange: DateRange) {
                state.startDate = dateRange.startDate;
                state.endDate = dateRange.endDate;
            },
            [UPDATE_CARDS_SELECTION](state: PaymentsState, id: string | null): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isSelected = !card.isSelected;

                        return card;
                    }

                    card.isSelected = false;

                    return card;
                });
            },
            [UPDATE_CARDS_DEFAULT](state: PaymentsState, id: string): void {
                state.creditCards = state.creditCards.map(card => {
                    if (card.id === id) {
                        card.isDefault = !card.isDefault;

                        return card;
                    }

                    card.isDefault = false;

                    return card;
                });
            },
            [SET_BILLING_HISTORY](state: PaymentsState, billingHistory: BillingHistoryItem[]): void {
                state.billingHistory = billingHistory;
            },
            [SET_PROJECT_USAGE_AND_CHARGES](state: PaymentsState, usageAndCharges: ProjectUsageAndCharges[]): void {
                state.usageAndCharges = usageAndCharges;
            },
            [SET_CURRENT_ROLLUP_PRICE](state: PaymentsState): void {
                state.currentRollupPrice = state.priceSummary;
            },
            [SET_PREVIOUS_ROLLUP_PRICE](state: PaymentsState): void {
                state.previousRollupPrice = state.priceSummary;
            },
            [SET_PRICE_SUMMARY](state: PaymentsState, charges: ProjectUsageAndCharges[]): void {
                if (charges.length === 0) {
                    state.priceSummary = 0;

                    return;
                }

                const usageItemSummaries = charges.map(item => item.summary());

                state.priceSummary = usageItemSummaries.reduce((accumulator, current) => accumulator + current);
            },
            [CLEAR](state: PaymentsState) {
                state.balance = 0;
                state.billingHistory = [];
                state.usageAndCharges = [];
                state.priceSummary = 0;
                state.previousRollupPrice = 0;
                state.currentRollupPrice = 0;
                state.creditCards = [];
                state.startDate = new Date();
                state.endDate = new Date();
            },
        },
        actions: {
            [GET_BALANCE]: async function({commit}: any): Promise<number> {
                const balance = await api.getBalance();

                commit(SET_BALANCE, balance);

                return balance;
            },
            [SETUP_ACCOUNT]: async function(): Promise<void> {
                await api.setupAccount();
            },
            [GET_CREDIT_CARDS]: async function({commit}: any): Promise<CreditCard[]> {
                const creditCards = await api.listCreditCards();

                commit(SET_CREDIT_CARDS, creditCards);

                return creditCards;
            },
            [ADD_CREDIT_CARD]: async function({commit}: any, token: string): Promise<void> {
                await api.addCreditCard(token);
            },
            [TOGGLE_CARD_SELECTION]: function({commit}: any, id: string): void {
                commit(UPDATE_CARDS_SELECTION, id);
            },
            [CLEAR_CARDS_SELECTION]: function({commit}: any): void {
                commit(UPDATE_CARDS_SELECTION, null);
            },
            [MAKE_CARD_DEFAULT]: async function({commit}: any, id: string): Promise<void> {
                await api.makeCreditCardDefault(id);

                commit(UPDATE_CARDS_DEFAULT, id);
            },
            [REMOVE_CARD]: async function({commit, state}: any, cardId: string): Promise<void> {
                await api.removeCreditCard(cardId);

                commit(SET_CREDIT_CARDS, state.creditCards.filter(card => card.id !== cardId));
            },
            [CLEAR_PAYMENT_INFO]: function({commit}: any): void {
                commit(CLEAR);
            },
            [GET_BILLING_HISTORY]: async function({commit}: any): Promise<void> {
                const billingHistory: BillingHistoryItem[] = await api.billingHistory();

                commit(SET_BILLING_HISTORY, billingHistory);
            },
            [MAKE_TOKEN_DEPOSIT]: async function({commit}: any, amount: number): Promise<TokenDeposit> {
                return await api.makeTokenDeposit(amount);
            },
            [GET_PROJECT_USAGE_AND_CHARGES]: async function({commit}: any, dateRange: DateRange): Promise<void> {
                const now = new Date();
                let beforeUTC = new Date(Date.UTC(dateRange.endDate.getUTCFullYear(), dateRange.endDate.getUTCMonth(), dateRange.endDate.getUTCDate(), 23, 59));

                if (now.getUTCFullYear() === dateRange.endDate.getUTCFullYear() &&
                    now.getUTCMonth() === dateRange.endDate.getUTCMonth() &&
                    now.getUTCDate() <= dateRange.endDate.getUTCDate()) {
                    beforeUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), now.getUTCHours(), now.getUTCMinutes()));
                }

                const sinceUTC = new Date(Date.UTC(dateRange.startDate.getUTCFullYear(), dateRange.startDate.getUTCMonth(), dateRange.startDate.getUTCDate(), 0, 0));
                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(sinceUTC, beforeUTC);

                commit(SET_DATE, dateRange);
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_CURRENT_ROLLUP]: async function({commit}: any): Promise<void> {
                const now = new Date();
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate(), now.getUTCHours(), now.getUTCMinutes()));
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1, 0, 0));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
                commit(SET_CURRENT_ROLLUP_PRICE);
            },
            [GET_PROJECT_USAGE_AND_CHARGES_PREVIOUS_ROLLUP]: async function({commit}: any): Promise<void> {
                const now = new Date();
                const startUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth() - 1, 1, 0, 0));
                const endUTC = new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 0, 23, 59, 59));

                const usageAndCharges: ProjectUsageAndCharges[] = await api.projectsUsageAndCharges(startUTC, endUTC);

                commit(SET_DATE, new DateRange(startUTC, endUTC));
                commit(SET_PROJECT_USAGE_AND_CHARGES, usageAndCharges);
                commit(SET_PRICE_SUMMARY, usageAndCharges);
                commit(SET_PREVIOUS_ROLLUP_PRICE);
            },
        },
        getters: {
            canUserCreateFirstProject: (state: PaymentsState): boolean => {
                return (state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && billingItem.status === BillingHistoryItemStatus.Completed;
                }) && state.balance > 0) || state.creditCards.length > 0;
            },
            isTransactionProcessing: (state: PaymentsState): boolean => {
                return state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && (billingItem.status === BillingHistoryItemStatus.Pending
                        || billingItem.status === BillingHistoryItemStatus.Paid
                        || billingItem.status === BillingHistoryItemStatus.Completed);
                }) && state.balance === 0;
            },
            isTransactionCompleted: (state: PaymentsState): boolean => {
                return (state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    return billingItem.amount >= 50 && billingItem.type === BillingHistoryItemType.Transaction
                        && billingItem.status === BillingHistoryItemStatus.Completed;
                }) && state.balance > 0);
            },
            isInvoiceForPreviousRollup: (state: PaymentsState): boolean => {
                const now = new Date();

                return state.billingHistory.some((billingItem: BillingHistoryItem) => {
                    if (now.getUTCMonth() === 0) {
                        return billingItem.type === BillingHistoryItemType.Invoice
                            && billingItem.start.getUTCFullYear() === now.getUTCFullYear() - 1
                            && billingItem.start.getUTCMonth() === 11;
                    }

                    return billingItem.type === BillingHistoryItemType.Invoice
                        && billingItem.start.getUTCFullYear() === now.getUTCFullYear()
                        && billingItem.start.getUTCMonth() === now.getUTCMonth() - 1;
                });
            },
        },
    };
}
