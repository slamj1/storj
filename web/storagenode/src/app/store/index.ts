// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vue from 'vue';
import Vuex from 'vuex';

import { makeNotificationsModule } from '@/app/store/modules/notifications';
import { makePayoutModule } from '@/app/store/modules/payout';
import { NotificationsHttpApi } from '@/storagenode/api/notifications';
import { PayoutHttpApi } from '@/storagenode/api/payout';
import { SNOApi } from '@/storagenode/api/storagenode';

import { appStateModule } from './modules/appState';
import { makeNodeModule } from './modules/node';

const notificationsApi = new NotificationsHttpApi();
const payoutApi = new PayoutHttpApi();
const nodeApi = new SNOApi();

Vue.use(Vuex);

/**
 * storage node store (vuex)
 */
export const store = new Vuex.Store({
   modules: {
       node: makeNodeModule(nodeApi),
       appStateModule,
       notificationsModule: makeNotificationsModule(notificationsApi),
       payoutModule: makePayoutModule(payoutApi),
   },
});

export default store;
