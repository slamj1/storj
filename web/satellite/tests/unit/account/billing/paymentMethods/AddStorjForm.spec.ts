// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import AddStorjForm from '@/components/account/billing/paymentMethods/AddStorjForm.vue';

import { appStateModule } from '@/store/modules/appState';
import { makeNotificationsModule } from '@/store/modules/notifications';
import { makePaymentsModule, PAYMENTS_MUTATIONS } from '@/store/modules/payments';
import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { makeUsersModule, USER_MUTATIONS } from '@/store/modules/users';
import { NOTIFICATION_MUTATIONS } from '@/store/mutationConstants';
import { BillingHistoryItem, BillingHistoryItemStatus, BillingHistoryItemType } from '@/types/payments';
import { Project } from '@/types/projects';
import { User } from '@/types/users';
import { Notificator } from '@/utils/plugins/notificator';
import { SegmentioPlugin } from '@/utils/plugins/segment';
import { createLocalVue, mount, shallowMount } from '@vue/test-utils';

import { PaymentsMock } from '../../../mock/api/payments';
import { ProjectsApiMock } from '../../../mock/api/projects';
import { UsersApiMock } from '../../../mock/api/users';

const localVue = createLocalVue();
const segmentioPlugin = new SegmentioPlugin();

localVue.use(Vuex);
localVue.use(segmentioPlugin);

const paymentsApi = new PaymentsMock();
const paymentsModule = makePaymentsModule(paymentsApi);
const usersApi = new UsersApiMock();
const usersModule = makeUsersModule(usersApi);
const projectsApi = new ProjectsApiMock();
const projectsModule = makeProjectsModule(projectsApi);
const notificationsModule = makeNotificationsModule();
const store = new Vuex.Store({ modules: { usersModule, paymentsModule, projectsModule, appStateModule, notificationsModule }});
store.commit(USER_MUTATIONS.SET_USER, new User('id', 'name', 'short', 'test@test.test', 'partner', 'pass'));

class NotificatorPlugin {
    public install() {
        localVue.prototype.$notify = new Notificator(store);
    }
}

const notificationsPlugin = new NotificatorPlugin();
localVue.use(notificationsPlugin);

describe('AddStorjForm', () => {
    it('renders correctly', () => {
        const wrapper = mount(AddStorjForm, {
            store,
            localVue,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('user is unable to add less than 50$ or more than 999999$ for the first time', async () => {
        const wrapper = mount(AddStorjForm, {
            store,
            localVue,
        });

        wrapper.vm.$data.tokenDepositValue = 30;
        await wrapper.vm.onConfirmAddSTORJ();

        expect((store.state as any).notificationsModule.notificationQueue[0].message).toMatch('First deposit amount must be more than $50 and less than $1000000');

        wrapper.vm.$data.tokenDepositValue = 1000000;
        await wrapper.vm.onConfirmAddSTORJ();

        expect((store.state as any).notificationsModule.notificationQueue[1].message).toMatch('First deposit amount must be more than $50 and less than $1000000');
    });

    it('user is able to add less than 50$ after coupon is applied', async () => {
        window.open = jest.fn();
        const billingTransactionItem = new BillingHistoryItem('itemId', 'test', 50, 50,
            BillingHistoryItemStatus.Completed, 'test', new Date(), new Date(), BillingHistoryItemType.Transaction);
        const project = new Project('testId', 'test', 'test', 'test', 'id', true);
        store.commit(NOTIFICATION_MUTATIONS.CLEAR);
        store.commit(PAYMENTS_MUTATIONS.SET_BILLING_HISTORY, [billingTransactionItem]);
        store.commit(PROJECTS_MUTATIONS.ADD, project);
        const wrapper = mount(AddStorjForm, {
            store,
            localVue,
        });

        wrapper.vm.$data.tokenDepositValue = 30;
        await wrapper.vm.onConfirmAddSTORJ();

        expect((store.state as any).notificationsModule.notificationQueue[0].type).toMatch('SUCCESS');
    });

    it('renders correctly after continue To Coin Payments click', () => {
        window.open = jest.fn();
        const wrapper = shallowMount(AddStorjForm, {
            store,
            localVue,
            propsData: {
                isLoading: true,
            },
        });

        expect(wrapper).toMatchSnapshot();
    });
});
