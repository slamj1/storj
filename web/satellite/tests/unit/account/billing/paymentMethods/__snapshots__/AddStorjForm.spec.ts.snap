// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStorjForm renders correctly 1`] = `
<div class="add-storj-area">
  <div class="add-storj-area__selection-container">
    <p class="add-storj-area__selection-container__label">Deposit STORJ Tokens via Coin Payments</p>
    <div class="form-container add-storj-area__selection-container__form">
      <div class="selected-container">
        <div id="paymentSelectButton" class="selected-container__label-container">
          <p class="selected-container__label-container__label">USD $50</p>
          <div class="selected-container__label-container__svg"><svg width="14" height="8" viewBox="0 0 14 8" fill="none" xmlns="http://www.w3.org/2000/svg">
              <path fill-rule="evenodd" clip-rule="evenodd" d="M0.372773 0.338888C0.869804 -0.112963 1.67565 -0.112963 2.17268 0.338888L7 4.72741L11.8273 0.338888C12.3243 -0.112963 13.1302 -0.112963 13.6272 0.338888C14.1243 0.790739 14.1243 1.52333 13.6272 1.97519L7 8L0.372773 1.97519C-0.124258 1.52333 -0.124258 0.790739 0.372773 0.338888Z" fill="#2683FF"></path>
            </svg></div>
        </div>
      </div>
      <!---->
      <!---->
      <!---->
    </div>
  </div>
  <div class="add-storj-area__submit-area">
    <!---->
    <div class="confirm-add-storj-button container" style="width: 251px; height: 48px;"><span class="label">Continue to Coin Payments</span></div>
  </div>
</div>
`;

exports[`AddStorjForm renders correctly after continue To Coin Payments click 1`] = `
<div class="add-storj-area">
  <div class="add-storj-area__selection-container">
    <p class="add-storj-area__selection-container__label">Deposit STORJ Tokens via Coin Payments</p>
    <tokendepositselection-stub class="add-storj-area__selection-container__form"></tokendepositselection-stub>
  </div>
  <div class="add-storj-area__submit-area"><img src="@/../static/images/account/billing/loading.gif" alt="loading gif" class="loading-image">
    <vbutton-stub label="Continue to Coin Payments" width="251px" height="48px" isdisabled="true" onpress="function () { [native code] }" class="confirm-add-storj-button"></vbutton-stub>
  </div>
</div>
`;
